sql/core/src/test/resources/sql-tests/results/postgreSQL/date.sql.out

```diff
@@ -848,7 +848,7 @@ SELECT DATE_TRUNC('CENTURY', TO_DATE('0055-08-10 BC', 'yyyy-MM-dd G'))
 -- !query schema
 struct<date_trunc(CENTURY, CAST(to_date('0055-08-10 BC', 'yyyy-MM-dd G') AS TIMESTAMP)):timestamp>
 -- !query output
--0099-01-01 00:00:00
+0100-01-01 00:00:00
```

Comment:
By constructing java.sql.Timestamp via date-time components, we loose opportunity to print negative years. Internally, the `Date` class performs normalization, and initializes the era but we don't print the era because we don't use `'G'` sub-pattern. 

If we specify `'G'` in our pattern, the era will printed for our current era as well. And this will break backward compatibility with previous Spark versions.

Reply:
Actually, the `toLocalDateTime()` of `java.sql.Timestamp` looses info about era:
```
    public LocalDateTime toLocalDateTime() {
        return LocalDateTime.of(getYear() + 1900,
                                getMonth() + 1,
                                getDate(),
                                getHours(),
                                getMinutes(),
                                getSeconds(),
                                getNanos());
```
`getYear()` returns 100 but it is the years in `BCE` era not in `CE`.

Reply:
I think this can be fixed by manually set the era in local date-time. Here is the example which works as expected:
```scala
def fromJavaTimestamp(t: Timestamp): SQLTimestamp = {
    if (t.getTime < GREGORIAN_CUTOVER_MILLIS) {
      val era = if (t.after(Timestamp.valueOf("0001-01-01 00:00:00"))) 1 else 0
      val ldt = t.toLocalDateTime.`with`(ChronoField.ERA, era)
      val instant = ZonedDateTime.of(ldt, ZoneId.systemDefault()).toInstant
      instantToMicros(instant)
    } else {
      instantToMicros(t.toInstant)
    }
  }
```
in the test:
```scala
    val formatter = TimestampFormatter.getFractionFormatter(ZoneOffset.UTC)
    val originLdt = LocalDateTime.of(-99, 1, 1, 0, 0, 0)
    val i = DateTimeUtils.instantToMicros(originLdt.atOffset(ZoneOffset.UTC).toInstant)
    val jts = DateTimeUtils.toJavaTimestamp(i)
    println(new SimpleDateFormat("yyyy G").format(jts))
    println(jts.toLocalDateTime)
    val back = DateTimeUtils.fromJavaTimestamp(jts)
    println(formatter.format(back))
```
```
0100 BC
0100-01-01T02:30:17
-0099-01-01 00:00:00
```

Reply:
Let's fix it!

Reply:
fixed